### PR TITLE
DbLedgerStorage implementation

### PR DESCRIPTION
**Pull request by merlimat**
Porting multiple squashed commits from Yahoo branch. 

This change adds a new `LedgerStorage` implementation that uses RocksDB to store the indexes for (ledgerId, entryId) to the data that is still kept in the entry logs.

In addition, there are also a `WriteCache` and a `ReadCache` that are used to completely decouple the read & write paths.

Note: This still has some methods that are not implemented. They are around the features added in 4.5 release: 

```java
public long getLastAddConfirmed(long ledgerId) throws IOException;
public Observable waitForLastAddConfirmedUpdate(long ledgerId, long previoisLAC, Observer observer) throws IOException;
public void setExplicitlac(long ledgerId, ByteBuf lac) throws IOException;
public ByteBuf getExplicitLac(long ledgerId);
```

These will be implemented in a subsequent commit.

### CODE REVIEW

File: bookkeeper-proto/src/main/proto/DbLedgerStorageDataFormats.proto [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+syntax = "proto2";
+
+option java_package = "org.apache.bookkeeper.bookie.storage.ldb";
+option optimize_for = SPEED;
+
+/**
+ * Ledger metadata stored in the bookie
+ */
+message LedgerData {
+	required bool exists = 1;
```
Comment (sijie) [inline bookkeeper-proto/src/main/proto/DbLedgerStorageDataFormats.proto]: tab => spaces

File: bookkeeper-server/pom.xml [hunk 2]
```diff
@@ -38,6 +38,11 @@
       <groupId>org.apache.bookkeeper</groupId>
       <artifactId>bookkeeper-proto</artifactId>
       <version>${project.parent.version}</version>
+	</dependency>
```
Comment (sijie) [inline bookkeeper-server/pom.xml]: tab => spaces

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java [hunk 3]
```diff
@@ -837,7 +840,7 @@ void flushRotatedLogs() throws IOException {
         leastUnflushedLogId = flushedLogId + 1;
     }
 
-    void flush() throws IOException {
+    public synchronized void flush() throws IOException {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: any reason why `synchronized` the whole method?

**Comment by sijie** [comment]
@merlimat I think there are findbugs and checkstyle errors. It is good if you can take a look and fix them. Since it is a new implementation, I only have a few comments on the places modifying existing code, otherwise, the change looks good to me.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            array[index + 5] = (byte) (value >>> 16);
+            array[index + 6] = (byte) (value >>> 8);
+            array[index + 7] = (byte) value;
+        }
+    }
+
+    public static final boolean isArrayAllZeros(final byte[] array) {
+        int sum = 0;
+        for (byte b : array) {
+            sum |= b;
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java]: Does this complete faster than early exiting with an if (b != 0) return false ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+                // Insert entry in read cache
+                readCache.put(orginalLedgerId, currentEntryId, entry);
+
+                count++;
+                size += entry.readableBytes();
+
+                currentEntryLocation += 4 + entry.readableBytes();
+                currentEntryLogId = currentEntryLocation >> 32;
+
+                entry.release();
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: This should be in a finally block?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+            }
+
+            readAheadBatchCountStats.registerSuccessfulValue(count);
+            readAheadBatchSizeStats.registerSuccessfulValue(size);
+        } catch (Exception e) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: Is there a motivation for catching Exception and not a more specific set of exception?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+            throw new IOException(e);
+        } finally {
+            isFlushOngoing.set(false);
+        }
+
+        return;
+    }
+
+    @Override
+    public synchronized void flush() throws IOException {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: There is a mix of synchronized blocks and usage of readwrite locks. I did not checked all the flows.
Just curious to know why don't you use only locks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndex.java [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+            }
+        } finally {
+            firstKeyWrapper.recycle();
+            lastKeyWrapper.recycle();
+            keyToDelete.recycle();
+            batch.close();
+        }
+    }
+
+    ///
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndex.java]: Empty comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/KeyValueStorageRocksDB.java [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+            try {
+                db.write(optionSync, writeBatch);
+            } catch (RocksDBException e) {
+                throw new IOException("Failed to flush RocksDB batch", e);
+            }
+        }
+    }
+
+    private static class EntryWrapper implements Entry<byte[], byte[]> {
+        private byte[] key;
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/KeyValueStorageRocksDB.java]: Final?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java (hunk 4 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java]: The theory is that removing the branching should get the optimizer job much easier, since it can then unroll the loop. 

This is only used to check for empty master key, so it won't be a long array in any case.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java (hunk 5 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: good point

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java (hunk 6 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: I think the intention was to ignore failures in read ahead, since we can already complete the current read entry operation.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java (hunk 7 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: Good point. That's more of a reflection of how the code evolved. This is actually the condensed view of 10s of iterations over the time. 

I think we can change the `synchronized` into a `ReentrantLock` to make it easier to follow.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java (hunk 3 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: There can only be a single `flush()` happening at a point in time. That will not affect the `addEntry()` which is only stopped while the 2 `WriteCache` instances are swapped.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        };
+
+        ArrayGroupSort sorter = new ArrayGroupSort(2, 4);
+        sorter.sort(data);
+
+        assertArrayEquals(expectedSorted, data);
+    }
+
+    @Test
+    public void errors() {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Maybe we can split in smaller methods? To make it clearer what the test does?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.fail;
+
+import org.junit.Test;
+
+public class ArraySortGroupTest {
+
+    @Test
+    public void simple() {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Don't know if it is better to add some test at least for most common corner cases like empty array, all equals values, odd/even number of elements.....

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ConversionRollbackTest.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Test
+    public void convertFromDbStorageToInterleaved() throws Exception {
+        File tmpDir = File.createTempFile("bkTest", ".dir");
+        tmpDir.delete();
+        tmpDir.mkdir();
+        File curDir = Bookie.getCurrentDirectory(tmpDir);
+        Bookie.checkDirectoryStructure(curDir);
+
+        System.out.println(tmpDir);
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ConversionRollbackTest.java]: Logger?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        }
+    }
+
+    public void deleteLedger(long ledgerId) {
+        deletedLedgers.add(ledgerId);
+    }
+
+    private static final ArrayGroupSort groupSorter = new ArrayGroupSort(2, 4);
+
+    public synchronized void forEach(EntryConsumer consumer) {
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Even if only one method is `synchronized`, still we need to have only 1 single thread calling the method at a time. 

The reason is that the `forEach()` iterates over the entries in order (by ledgerId, entryId), thus it needs to convert the hashmap into a sorted array of pair of longs. The array used is a member variable of the instance (`long[] sortedEntries`).

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java [hunk 14]
```diff
@@ -0,0 +1,767 @@
...
+            WriteCache tmp = writeCacheBeingFlushed;
+            writeCacheBeingFlushed = writeCache;
+            writeCache = tmp;
+
+            // since the cache is switched, we can allow flush to be triggered
+            hasFlushBeenTriggered.set(false);
+            flushWriteCacheCondition.signalAll();
+        } finally {
+            isFlushOngoing.set(true);
+            writeCacheMutex.writeLock().unlock();
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: Well, the only reason for `writeCacheMutex` to not be unlocked would be to have `isFlushOngoing == null`, but its a final Object assigned in the constructor.

For the method logic, it's critical to set the flag, to mark an ongoing flush before releasing the write lock. That is used when a flush is triggered by the write cache being full. In this case we want to trigger 1 (and only 1) flush and wait until the write cache is rotated to put the entry in the new empty write cache and return.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    }
+
+    @Override
+    public void close() {
+        cacheSegments.forEach(ByteBuf::release);
+    }
+
+    public void put(long ledgerId, long entryId, ByteBuf entry) {
+        lock.readLock().lock();
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java]: Sure, the 2 lines there don't need to be in the read lock section.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java (hunk 11 above)
Comment (merlimat) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Sure, added a bit more of these

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java [hunk 16]
```diff
@@ -0,0 +1,767 @@
...
+            readCache.put(ledgerId, entryId, entry);
+
+            // Try to read more entries
+            long nextEntryLocation = entryLocation + 4 /* size header */ + entry.readableBytes();
+            fillReadAheadCache(ledgerId, entryId + 1, nextEntryLocation);
+
+            recordSuccessfulEvent(readCacheMissStats, startTime);
+            recordSuccessfulEvent(readEntryStats, startTime);
+            return entry;
+        } catch (NoEntryException e) {
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: Sure, makes sense. `NoEntryException` cannot be thrown after that point.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java]: Actually, findbugs catches it. Just resolved to additional `try / finally` block to make findbugs happy :)

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java (hunk 3 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: Uhm, I thought this was `LedgerStorage.flush()`. Let me revisit this.

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: Yes, that was not needed indeed.

**Comment by merlimat** [comment]
Addressed comments & fixed checkstyle and findbugs warnings

**Comment by merlimat** [comment]
retest this please

**Comment by merlimat** [comment]
@jiazhai eventually I ended up doing similar change and that was merged in https://github.com/yahoo/bookkeeper/commit/c4f1f9aca8ea9a8d470626e220fb9d3d06cdcafb

This change is what the Yahoo branch was having and its included in this PR.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java (hunk 4 above)
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java]: FYI: I have double checked, PlatformDependant has a "isZero" method

**Comment by jiazhai** [comment]
👍, I C, right it is included.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java (hunk 4 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ArrayUtil.java]: 👍  Nice. I think that was added recently. Will switch to that.

**Comment by merlimat** [comment]
@ivankelly Sure, I can break it up that way.

**Comment by merlimat** [comment]
I think I have addressed all comments so far. Closing it right now to send out individual patches.
